include/xtensor/xfft.hpp

```diff
@@ -0,0 +1,317 @@
+
+                // Temporary vectors and preprocessing
+                xt::xarray<std::complex<T>> av = xt::empty<std::complex<T>>({ m });
+                xt::view(av, xt::range(0, n)) = data * expTable;
+
+
+                xt::xarray<std::complex<T>> bv = xt::empty<std::complex<T>>({ m });
+                xt::view(bv, xt::range(0, n)) = xt::conj(expTable);
+                xt::view(bv, xt::range(-n + 1, xt::placeholders::_)) = xt::view(xt::conj(xt::flip(expTable)), xt::range(xt::placeholders::_, - 1));
+
+                // Convolution
+                auto cv = xt::fft::fft_convolve<T>(av, bv);
+
+                return xt::eval(xt::view(cv, xt::range(0, n)) * expTable);
+            }
+
+            template<
+                typename T
+            >
+            auto fft_impl(xt::xarray<std::complex<T>> data)
+            {
+                //loop through all the next highest axis
+
+                // Is power of 2
+                if (detail::is_power2(data))
+                {
+                    return detail::transform_radix2<T>(std::move(data));
+                }
+                //More complicated algorithm for arbitrary sizes
+                else 
+                {
+                    return detail::transform_bluestein<T>(std::move(data));
+                }
+            }
+
+            template<
+                typename T, 
+                typename E1
+            >
+            auto fft(E1&& data, std::ptrdiff_t axis = -1)
+            {
+                //check the length of the data on that axis
+                const std::size_t n(data.shape(axis));
+                if (n == 0) {
+                    XTENSOR_THROW(std::runtime_error, "Cannot take the FFT along an empty dimention");
+                }
+
+                //select the axis
+                //create the return type
+                //this can be made smarter to use floats or double for small speed up
+                xt::xarray<std::complex<T>> complex_out = xt::zeros<std::complex<T>>(data.shape());
+                auto saxis = xt::normalize_axis(data.dimension(), axis);
+                auto begin = xt::axis_slice_begin(data, saxis);
+                auto end = xt::axis_slice_end(data, saxis);
+                auto iter_out = xt::axis_slice_begin(complex_out, saxis);
+
+                for (auto iter = begin; iter != end; iter++)
+                {
+                    (*iter_out++) = detail::fft_impl<T>(*iter);
+                }
+
+                return complex_out;
+            }
+        }
+
+        //meta class for finding nested types inside complex types
+        //false case
+        template<typename T>
+        struct is_complex_t : public std::false_type 
+        {
+        public:
+            using value_type = T;
+        };
+        
+        //true case
+        template<typename T>
+        struct is_complex_t<std::complex<T>> : public std::true_type 
+        {
+        public:
+            using value_type = T;
+        };
```

Comment:
There might be simpler ways, see e.g. [here](https://coder-question.com/cq-blog/107822). And probably `xtl` can help, @JohanMabille  ? Anyways, this should not be part of `xtensor`'s public API

Reply:
We provide this metafunction in the [xtl](https://github.com/xtensor-stack/xtl/blob/master/include/xtl/xcomplex.hpp#L57). It handles references and the `xcomplex` class too, so it might be worth using it.